analyzers/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldContainAssertion.cs

```diff
@@ -98,6 +95,36 @@ protected override void Initialize(SonarAnalysisContext context)
                 SyntaxKind.MethodDeclaration);
         }
 
+        private static bool ContainsAssertion(MethodDeclarationSyntax methodDeclaration, SemanticModel previousSemanticModel, ISet<IMethodSymbol> visitedSymbols, int level)
+        {
+            var currentSemanticModel = methodDeclaration.EnsureCorrectSemanticModel(previousSemanticModel);
+            var invokedSymbols = methodDeclaration.DescendantNodes()
+                .OfType<InvocationExpressionSyntax>()
+                .Select(expression => currentSemanticModel.GetSymbolInfo(expression).Symbol)
+                .OfType<IMethodSymbol>();
+
+            if (invokedSymbols.Any(symbol => IsKnownAssertion(symbol) || IsCustomAssertion(symbol)))
+            {
+                return true;
+            }
+            if (level == MaxInvocationDepth)
+            {
+                return false;
+            }
+            foreach (var symbol in invokedSymbols.Where(x => !visitedSymbols.Contains(x)))
+            {
+                visitedSymbols.Add(symbol);
+                foreach (var invokedDeclaration in symbol.DeclaringSyntaxReferences.Select(x => x.GetSyntax()).OfType<MethodDeclarationSyntax>())
+                {
+                    if (ContainsAssertion(invokedDeclaration, currentSemanticModel, visitedSymbols, level + 1))
```

Comment:
If I remember correctly we talked about BFS implementation and this one looks like DFS. Or am I missing something?

Reply:
I've skipped the BFS idea to keep it simple, because I've chosen the limit to be 2 instead of the original idea of 3